Trasnform payload lists into tables (#4158)
telefonicaid/fiware-orion: doc/manuals/orion-api.md

            - [List Entity Types [GET /v2/type]](#list-entity-types-get-v2type)
            - [Retrieve entity information for a given type [GET /v2/types]](#retrieve-entity-information-for-a-given-type-get-v2types)
    - [Subscriptions Operations](#subscriptions-operations)

>         - [Subscription payload datamodel](#Subscriptionpayloaddatamodel)

Reviewer: ```suggestion
        - [Subscription payload datamodel](#subscriptionpayloaddatamodel)
```

Author: done in https://github.com/telefonicaid/fiware-orion/pull/4158/commits/88575a0a00f9e89f0fa45e273506a4186183febb

            - [Update Subscription [PATCH /v2/subscriptions/{subscriptionId}]](#update-subscription-patch-v2subscriptionssubscriptionid)
            - [Delete subscription [DELETE /v2/subscriptions/{subscriptionId}]](#delete-subscription-delete-v2subscriptionssubscriptionid)
    - [Registration Operations](#registration-operations)

>         - [Registration payload datamodel](#Registrationpayloaddatamodel)

Reviewer: ```suggestion
        - [Registration payload datamodel](#registrationpayloaddatamodel)
```

> A subscription is represented by a JSON object with the following fields:

Author: Done in https://github.com/telefonicaid/fiware-orion/pull/4158/commits/11a38fa376766033e684811ddff1085f0b73a8aa

> A subscription is represented by a JSON object with the following fields:

Author: Please review titles name

| `throttling`   | ✓        | number | Minimal period of time in seconds which must elapse between two consecutive notifications.    |

A `subject` contains the following subfields:

| Parameter      | Optional | Type   | Description                                                                                   |
|----------------|----------|--------|-----------------------------------------------------------------------------------------------|
| `entities`     | ✓        | string | A list of objects, each one composed of the following subfields: <ul><li><code>id</code> or <code>idPattern</code> Id or pattern of the affected entities. Both cannot be used at the same time, but one of them must be present.</li> <li><code>type</code> or <code>typePattern</code> Type or type pattern of the affected entities. Both cannot be used at the same time. If omitted, it means "any entity type".</li></ul> |

> | `condition`    | ✓        | string | Condition to trigger notifications. This field is optional and it may contain two properties, both optional: <ul><li><code>attrs</code> Array of attribute names that will trigger the notification. </li> <li><code>expression</code> An expression composed of <code>q</code>, <code>mq</code>, <code>georel</code>, <code>geometry</code> and <code>coords</code> (see "List entities" operation above about this field)</li></ul> |

Reviewer: This is an one-line-HTML for a bullet list, but I'm my opinion is "under control"

Maybe that's the criteria to choose if a given level has to be "sub-tabled" or not:

* If the HTML is too complex/awful, then it deserves a new sub-table to be linked from here
* If the HTML is manageable (as this one) an description within the cell itself is enough.

Reviewer: NTC

| `http` or `httpCustom`   | ✓        | object | One of them must be present, but not both at the same time. It is used to convey parameters for notifications delivered through the HTTP protocol. |
| `attrsFormat`            | ✓        | string | Specifies how the entities are represented in notifications. Accepted values are `normalized` (default), `keyValues` or `values`.<br> If `attrsFormat` takes any value different than those, an error is raised. See detail in "Notification Messages" section. |
| `metadata`               | ✓        | string | List of metadata to be included in notification messages. See "Filtering out attributes and metadata" section for more detail. |

> | `timesSent`              |          | number | Not editable, only present in GET operations. Number of notifications sent due to this subscription. |
> | `lastNotification`       |          | string | Not editable, only present in GET operations. Last notification timestamp in ISO8601 format. |
> | `lastFailure`            |          | string | Not editable, only present in GET operations. Last failure timestamp in ISO8601 format. Not present if subscription has never had a problem with notifications. |
> | `lastSuccess`            |          | string | Not editable, only present in GET operations. Timestamp in ISO8601 format for last successful notification.  Not present if subscription has never had a successful notification. |

Reviewer: We should some way explain that some parameters cannot be used when creating or updating a subscription as they are only used in GET responses to some some status/statistical information.

One solution could be to use the "Optional?" column this way

```suggestion
| `timesSent`              | Only in GET         | number | Not editable, only present in GET operations. Number of notifications sent due to this subscription. |
| `lastNotification`       | Only in GET          | string | Not editable, only present in GET operations. Last notification timestamp in ISO8601 format. |
| `lastFailure`            |  Only in GET        | string | Not editable, only present in GET operations. Last failure timestamp in ISO8601 format. Not present if subscription has never had a problem with notifications. |
| `lastSuccess`            | Only in GET         | string | Not editable, only present in GET operations. Timestamp in ISO8601 format for last successful notification.  Not present if subscription has never had a successful notification. |
```

Although maybe another alternatives are also possible.

Reviewer: With regards to optional fields used in creation and updating requests, we have this cases:

1. Fields that are not provided and in the GET responses are not shown
2. Fields that are not provided and in the GET responses use a default value

(I think we started implementing according to style 1, but newer additions, e.g. `onlyChangedAttrs`, use style 2)

Should be distinguish both cases in the "Optional" column some way? Unsure...

Reviewer: "Only on retrieval"

Reviewer: "GET responses" -> "Only on retrieval"

Author: Done in https://github.com/telefonicaid/fiware-orion/pull/4158#:~:text=to%20optional%20col-,f24e36b,-mapedraza%20force%2Dpushed

## Subscriptions Operations

### Subscription payload datamodel

#### `subscription`

A subscription is represented by a JSON object with the following fields:

| Parameter      | Optional | Type   | Description                                                                                   |
|----------------|----------|--------|-----------------------------------------------------------------------------------------------|
| `id`           |          | string | Subscription unique identifier. Automatically created at creation time.                       |
| `description`  | ✓        | string | A free text used by the client to describe the subscription.                                  |

> | `subject`      |          | object | An object that describes the subject of the subscription.                                     |

Reviewer: ```suggestion
| [`subject`](#subscriptionsubject)      |          | object | An object that describes the subject of the subscription.                                     |
```

| [`notification`](#subscriptionnotification) |          | object | An object that describes the notification to send when the subscription is triggered.         |

> | `expires`      | ✓        | string | Subscription expiration date in ISO8601 format. Permanent subscriptions must omit this field. |

Reviewer: Unsure... maybe it's redundant.

| `status`       |          | string | Either `active` (for active subscriptions) or `inactive` (for inactive subscriptions). If this field is not provided at subscription creation time, new subscriptions are created with the `active` status, which can be changed by clients afterwards. For expired subscriptions, this attribute is set to `expired` (no matter if the client updates it to `active`/`inactive`). Also, for subscriptions experiencing problems with notifications, the status is set to `failed`. As soon as the notifications start working again, the status is changed back to `active`.                                                                                              |
| `throttling`   | ✓        | number | Minimal period of time in seconds which must elapse between two consecutive notifications.    |

#### `subscription.subject`

A `subject` contains the following subfields:

| Parameter      | Optional | Type   | Description                                                                                   |
|----------------|----------|--------|-----------------------------------------------------------------------------------------------|

> | `entities`     | ✓        | string | A list of objects, each one composed of the following subfields: <ul><li><code>id</code> or <code>idPattern</code> Id or pattern of the affected entities. Both cannot be used at the same time, but one of them must be present.</li> <li><code>type</code> or <code>typePattern</code> Type or type pattern of the affected entities. Both cannot be used at the same time. If omitted, it means "any entity type".</li></ul> |
> | `condition`    | ✓        | string | Condition to trigger notifications. This field is optional and it may contain two properties, both optional: <ul><li><code>attrs</code> Array of attribute names that will trigger the notification. </li> <li><code>expression</code> An expression composed of <code>q</code>, <code>mq</code>, <code>georel</code>, <code>geometry</code> and <code>coords</code> (see "List entities" operation above about this field)</li></ul> |

Reviewer: ```suggestion
| `entities`     | ✓        | array| A list of objects, each one composed of the following subfields: <ul><li><code>id</code> or <code>idPattern</code> Id or pattern of the affected entities. Both cannot be used at the same time, but one of them must be present.</li> <li><code>type</code> or <code>typePattern</code> Type or type pattern of the affected entities. Both cannot be used at the same time. If omitted, it means "any entity type".</li></ul> |
| `condition`    | ✓        | object| Condition to trigger notifications. This field is optional and it may contain two properties, both optional: <ul><li><code>attrs</code> Array of attribute names that will trigger the notification. </li> <li><code>expression</code> An expression composed of <code>q</code>, <code>mq</code>, <code>georel</code>, <code>geometry</code> and <code>coords</code> (see "List entities" operation above about this field)</li></ul> |
```

#### `subscription.notification`

A `notification` object contains the following subfields:

| Parameter              | Optional          | Type   | Description                                                                                                                                                                                                                                                     |
|------------------------|-------------------|--------|-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|

> | `attrs` or `exceptAttrs` |          | string | Both cannot be used at the same time. <ul><li><code>attrs</code>: List of attributes to be included in notification messages. It also defines the order in which attributes must appear in notifications when <code>attrsFormat</code> <code>value</code> is used (see "Notification Messages" section). An empty list means that all attributes are to be included in notifications. See "Filtering out attributes and metadata" section for more detail.</li><li><code>exceptAttrs</code>: List of attributes to be excluded from the notification message, i.e. a notification message includes all entity attributes except the ones listed in this field.</li><li>If neither <code>attrs</code> nor <code>exceptAttrs</code> is specified, all attributes are included in notifications.</li></ul>|

Reviewer: ```suggestion
| `attrs` or `exceptAttrs` |          | array | Both cannot be used at the same time. <ul><li><code>attrs</code>: List of attributes to be included in notification messages. It also defines the order in which attributes must appear in notifications when <code>attrsFormat</code> <code>value</code> is used (see "Notification Messages" section). An empty list means that all attributes are to be included in notifications. See "Filtering out attributes and metadata" section for more detail.</li><li><code>exceptAttrs</code>: List of attributes to be excluded from the notification message, i.e. a notification message includes all entity attributes except the ones listed in this field.</li><li>If neither <code>attrs</code> nor <code>exceptAttrs</code> is specified, all attributes are included in notifications.</li></ul>|
```

| `attrsFormat`          | ✓                 | string | Specifies how the entities are represented in notifications. Accepted values are `normalized` (default), `keyValues` or `values`.<br> If `attrsFormat` takes any value different than those, an error is raised. See detail in "Notification Messages" section. |
| `metadata`             | ✓                 | string | List of metadata to be included in notification messages. See "Filtering out attributes and metadata" section for more detail.                                                                                                                                  |
| `timesSent`            | Only on retrieval | number | Not editable, only present in GET operations. Number of notifications sent due to this subscription.                                                                                                                                                            |

> | `lastNotification`     | Only on retrieval | string | Not editable, only present in GET operations. Last notification timestamp in ISO8601 format.                                                                                                                                                                    |
> | `lastFailure`          | Only on retrieval | string | Not editable, only present in GET operations. Last failure timestamp in ISO8601 format. Not present if subscription has never had a problem with notifications.                                                                                                 |
> | `lastSuccess`          | Only on retrieval | string | Not editable, only present in GET operations. Timestamp in ISO8601 format for last successful notification.  Not present if subscription has never had a successful notification.                                                                               |

Reviewer: ```suggestion
| `lastNotification`     | Only on retrieval | ISO 8601 string | Not editable, only present in GET operations. Last notification timestamp in ISO8601 format.                                                                                                                                                                    |
| `lastFailure`          | Only on retrieval | ISO 8601 string | Not editable, only present in GET operations. Last failure timestamp in ISO8601 format. Not present if subscription has never had a problem with notifications.                                                                                                 |
| `lastSuccess`          | Only on retrieval | ISO 8601 string | Not editable, only present in GET operations. Timestamp in ISO8601 format for last successful notification.  Not present if subscription has never had a successful notification.                                                                               |
```

#### `subscription.notification.http`

An `http` object contains the following subfields:

| Parameter | Optional | Type   | Description                                                                                   |
|-----------|----------|--------|-----------------------------------------------------------------------------------------------|
| `url`     |          | string | URL referencing the service to be invoked when a notification is generated. An NGSIv2 compliant server must support the `http` URL schema. Other schemas could also be supported. |

#### `subscription.notification.httpCustom`

An `httpCustom` object contains the following subfields.

| Parameter | Optional | Type   | Description                                                                                   |
|-----------|----------|--------|-----------------------------------------------------------------------------------------------|
| `url`     |          | string | Same as in `http` above.                                                                      |

> | `headers` | ✓        | string | A key-map of HTTP headers that are included in notification messages.                         |
> | `qs`      | ✓        | string | A key-map of URL query parameters that are included in notification messages.                 |

Reviewer: ```suggestion
| `headers` | ✓        | object | A key-map of HTTP headers that are included in notification messages.                         |
| `qs`      | ✓        | object | A key-map of URL query parameters that are included in notification messages.                 |
```

| [`condition`](#subscriptionsubjectcondition) | ✓        | object | Condition to trigger notifications. If omitted, it means "any attribute change will trigger condition" |

#### `subscription.subject.condition`

A `condition` contains the following subfields:

| Parameter    | Optional | Type  | Description                                                                                                                   |
|--------------|----------|-------|-------------------------------------------------------------------------------------------------------------------------------|
| `attrs`      | ✓        | array | Array of attribute names that will trigger the notification.                                                                  |
| `expression` | ✓        | object| An expression composed of `q`, `mq`, `georel`, `geometry` and `coords` (see "List entities" operation above about this field) |

> `condition` field modifies the triggering Notification rules as follow:

Reviewer: ```suggestion
Based on the `condition` field, the notification triggering rules are as follow:
```

| [`dataProvided`](#registrationdataprovided)          |          | object | Object that describes the data provided by this source.                                                                                                                                     |
| `status`       | ✓        | string | Enumerated field which captures the current status of this registration with the possibles values: [`active`, `inactive`, `expired` or `failed`]. If this field is not provided at registration creation time, new registrations are created with the `active` status, which may be changed by clients afterwards. For expired registrations, this attribute is set to `expired` (no matter if the client updates it to `active`/`inactive`). Also, for registrations experiencing problems with forwarding operations, the status is set to `failed`. As soon as the forwarding operations start working again, the status is changed back to `active`. |

> | `expires`               | ✓        | string | Registration expiration date in ISO8601 format. Permanent registrations must omit this field.                                                                                               |

Reviewer: ```suggestion
| `expires`               | ✓        | ISO 8601 string | Registration expiration date in ISO8601 format. Permanent registrations must omit this field.                                                                                               |
```

| [`forwardingInformation`](#registrationforwardinginformation) |          | object | Information related to the forwarding operations made against the provider. Automatically provided by the implementation, in the case such implementation supports forwarding capabilities. |

#### `registration.provider`

The `provider` field contains the following subfields:

| Parameter      | Optional | Type   | Description                                                                                   |
|----------------|----------|--------|-----------------------------------------------------------------------------------------------|

> | `http`         |          | string | It is used to convey parameters for providers that deliver information through the HTTP protocol.(Only protocol supported nowadays). <br>It must contain a subfield named `url` with the URL that serves as the endpoint that offers the providing interface. The endpoint must *not* include the protocol specific part (for instance `/v2/entities`). |

Reviewer: ```suggestion
| `http`         |          | object | It is used to convey parameters for providers that deliver information through the HTTP protocol.(Only protocol supported nowadays). <br>It must contain a subfield named `url` with the URL that serves as the endpoint that offers the providing interface. The endpoint must *not* include the protocol specific part (for instance `/v2/entities`). |
```

| `supportedForwardingMode`  |          | string | It is used to convey the forwarding mode supported by this context provider. By default `all`. Allowed values are: <ul><li><code>none</code>: This provider does not support request forwarding.</li><li><code>query</code>: This provider only supports request forwarding to query data.</li><li><code>update</code>: This provider only supports request forwarding to update data.</li><li><code>all</code>: This provider supports both query and update forwarding requests. (Default value).</li></ul> |

#### `registration.dataProvided`

The `dataProvided` field contains the following subfields:

| Parameter      | Optional | Type   | Description                                                                                   |
|----------------|----------|--------|-----------------------------------------------------------------------------------------------|

> | `entities`     |          | string | A list of objects, each one composed of the following subfields: <ul><li><code>id</code> or <code>idPattern</code>: d or pattern of the affected entities. Both cannot be used at the same time, but one of them must be present.</li><li><code>type</code> or <code>typePattern</code>: Type or pattern of the affected entities. Both cannot be used at the same time. If omitted, it means "any entity type".</li></ul> |
> | `attrs`        |          | string | List of attributes to be provided (if not specified, all attributes). |
> | `expression`   |          | string | By means of a filtering expression, allows to express what is the scope of the data provided. Currently only geographical scopes are supported through the following subterms: <ul><li><code>georel</code>: Any of the geographical relationships as specified by the Geoqueries section of this specification. </li><li><code>geometry</code>: Any of the supported geometries as specified by the Geoqueries section of this specification.</li> <li><code>coords</code>: String representation of coordinates as specified by the Geoqueries section of this specification.</li></ul> |

Reviewer: ```suggestion
| `entities`     |          | array | A list of objects, each one composed of the following subfields: <ul><li><code>id</code> or <code>idPattern</code>: d or pattern of the affected entities. Both cannot be used at the same time, but one of them must be present.</li><li><code>type</code> or <code>typePattern</code>: Type or pattern of the affected entities. Both cannot be used at the same time. If omitted, it means "any entity type".</li></ul> |
| `attrs`        |          | array | List of attributes to be provided (if not specified, all attributes). |
| `expression`   |          | object | By means of a filtering expression, allows to express what is the scope of the data provided. Currently only geographical scopes are supported through the following subterms: <ul><li><code>georel</code>: Any of the geographical relationships as specified by the Geoqueries section of this specification. </li><li><code>geometry</code>: Any of the supported geometries as specified by the Geoqueries section of this specification.</li> <li><code>coords</code>: String representation of coordinates as specified by the Geoqueries section of this specification.</li></ul> |
```

#### `registration.forwardingInformation`

The `forwardingInformation` field contains the following subfields:

| Parameter        | Optional          | Type   | Description                                                                                                                                                                            |
|------------------|-------------------|--------|----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|

> | `timesSent`      | Only on retrieval | string | Not editable, only present in GET operations. Number of request forwardings sent due to this registration.                                                                             |

Reviewer: ```suggestion
| `timesSent`      | Only on retrieval | number | Not editable, only present in GET operations. Number of request forwardings sent due to this registration.                                                                             |
```

> | `lastForwarding` | Only on retrieval | string | Not editable, only present in GET operations. Last forwarding timestamp in ISO8601 format.                                                                                             |
> | `lastFailure`    | Only on retrieval | string | Not editable, only present in GET operations. Last failure timestamp in ISO8601 format. Not present if registration has never had a problem with forwarding.                           |
> | `lastSuccess`    | Only on retrieval | string | Not editable, only present in GET operations. Timestamp in ISO8601 format for last successful request forwarding. Not present if registration has never had a successful notification. |

Reviewer: ```suggestion
| `lastForwarding` | Only on retrieval | ISO 8601 string | Not editable, only present in GET operations. Last forwarding timestamp in ISO8601 format.                                                                                             |
| `lastFailure`    | Only on retrieval | ISO 8601 string | Not editable, only present in GET operations. Last failure timestamp in ISO8601 format. Not present if registration has never had a problem with forwarding.                           |
| `lastSuccess`    | Only on retrieval | ISO 8601 string | Not editable, only present in GET operations. Timestamp in ISO8601 format for last successful request forwarding. Not present if registration has never had a successful notification. |
```
